File: C++/Step12-Dynamic Programming_1/2579.cpp
//
//  2579.cpp
//  Step12-Dynamic Programming_1
//
//  Created by 김예빈 on 2019. 9. 10..
//  Copyright © 2019년 김예빈. All rights reserved.
//

#include <iostream>
using namespace std;

int main(int argc, const char * argv[]) {
    cin.tie(NULL);
    ios::sync_with_stdio(false);
    
    int n, i;
    
    cin >> n;
    
    int step[n];
    int dp[n];
    
    for(i = 0; i < n; i++)
        cin >> step[i];
    
    dp[0] = step[0];
    dp[1] = step[0] + step[1] > step[1] ? step[0] + step[1] : step[1];
    dp[2] = step[0] + step[2] > step[1] + step[2] ? step[0] + step[2] : step[1] + step[2];
    
    for(i = 3; i < n; i++)
        dp[i] = (dp[i - 3] + step[i - 1] > dp[i - 2] ? dp[i - 3] + step[i - 1] : dp[i - 2]) + step[i];
    
    cout << dp[n - 1];
    
    return 0;
}

File: BaekJoon Online Judge/Swift/Step14-Dynamic Programming_1/2579.swift
//
//  2579.swift
//  Step14-Dynamic Programming_1
//
//  Created by Yebin Kim on 2019/12/15.
//  Copyright © 2019 김예빈. All rights reserved.
//

import Foundation

if let N = Int(readLine()!) {
    var step = Array<Int>(repeating: 0, count: 1)
    var dp = Array(repeating: 0, count: N + 1)
    
    for _ in 0..<N {
        step.append(Int(readLine()!)!)
    }
    
    dp[1] = step[1]
    dp[2] = step[1] + step[2]
    
    for i in 3..<dp.count {
        dp[i] = max(dp[i - 2], step[i - 1] + dp[i - 3])
        dp[i] += step[i]
    }
    
    print(dp.last!)
}
